Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -140,82 +152,72 @@ public actor KafkaProducer {
 
         let client = try RDKafka.createClient(
             type: .producer,
             configDictionary: config.dictionary,
-            deliveryReportCallback: { [logger, streamContinuation] messageResult in
+            deliveryReportCallback: { [logger, source] messageResult in
                 guard let messageResult else {
                     logger.error("Could not resolve acknowledged message")
                     return
                 }
 
                 // Ignore YieldResult as we don't support back pressure in KafkaProducer
-                streamContinuation?.yield(messageResult)
+                _ = source.yield(messageResult)
             },
             logger: logger
         )
 
-        let producer = try await KafkaProducer(
-            client: client,
+        let producer = try KafkaProducer(
+            stateMachine: stateMachine,
             config: config,
-            topicConfig: topicConfig,
-            logger: logger
+            topicConfig: topicConfig
         )
 
-        streamContinuation?.onTermination = { [producer] _ in
-            Task {
-                await producer.shutdownGracefully()
-            }
+        stateMachine.withLockedValue {
+            $0.initialize(
+                client: client,
+                source: source
+            )
         }
 
-        let acknowlegementsSequence = KafkaMessageAcknowledgements(wrappedSequence: stream)
+        let acknowlegementsSequence = KafkaMessageAcknowledgements(wrappedSequence: sourceAndSequence.sequence)
         return (producer, acknowlegementsSequence)
     }
 
     /// Method to shutdown the ``KafkaProducer``.
     ///
     /// This method flushes any buffered messages and waits until a callback is received for all of them.
     /// Afterwards, it shuts down the connection to Kafka and cleans any remaining state up.
-    /// - Parameter timeout: Maximum amount of milliseconds this method waits for any outstanding messages to be sent.
-    public func shutdownGracefully(timeout: Int32 = 10000) async {
-        switch self.state {
-        case .started:
-            self.state = .shuttingDown
-            await self._shutdownGracefully(timeout: timeout)
-        case .shuttingDown, .shutDown:
-            return
-        }
-    }
-
-    private func _shutdownGracefully(timeout: Int32) async {
-        await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
-            // Wait `timeout` seconds for outstanding messages to be sent and callbacks to be called
-            self.client.withKafkaHandlePointer { handle in
-                rd_kafka_flush(handle, timeout)
-                continuation.resume()
-            }
-        }
-
-        for (_, topicHandle) in self.topicHandles {
-            rd_kafka_topic_destroy(topicHandle)
-        }
-
-        self.state = .shutDown
+    public func shutdownGracefully() {
+        self.stateMachine.withLockedValue { $0.finish() }
```

Comment:
Let's call this `triggerGracefulShutdown` since we are actually not waiting for the shutdown here. Once we adopt lifecycle we should make this method non `public`